Add Theme Showcase Banner

mendezcode opened the pull request:
Adds a banner component to Calypso. The banner is intended to be used initially on the theme showcase, as it's shown in the following screenshot:

![screen shot 2017-11-23 at 5 42 35 pm](https://user-images.githubusercontent.com/5881557/33189073-bc4bc5bc-d075-11e7-8df5-2589109a3d74.png)

The Banner component is `<ThemesBanner />`, and it receives `items`, which is an array of objects, each with this signature:

```js
{
  title,
  description,
  buttonLabel,
  buttonType,
  buttonAction,
  backgroundImage,
  backgroundColor,
  textColor,
  textShadowColor,
}
```

_NOTE: This is a rough PR of the actual implementation. There are still some things to take care of, which will be sorted out after we go over the first round of review. More info on the NOTES section._

- - -

### Testing Instructions

Add something like this on the `render()` method of `theme-showcase.jsx`:

```js
const bannerItems = [
	{
		title: "Are you a small business?",
		description: "We understand your time contraints, and that's why we created this easy-to-set-up Premium theme called Small Business.",
		buttonLabel: "Learn about this theme",
		buttonType: "primary",
		backgroundImage: "/calypso/images/theme-banners/banner-bg.jpg",
		buttonAction: this.helloWorld,
	}
];
```

Then, add the following JSX above `<ThemesSearchCard ... />`:

```jsx
<ThemesBanner items={ bannerItems } />
```

The image used is [here](https://user-images.githubusercontent.com/5881557/33189130-55035d7e-d076-11e7-9cad-6283e87cedbf.jpg).

### NOTES

Not sure how this will be handled in detail, but maybe for v1 we can show the banner items randomly on load, or we can add a scrolling mechanism that scrolls the banner items dynamically too.

Right now, banner items will be rendered one on top of the other, until a proper behavior has been discussed here. CC @mapk 

- - -

Any design suggestions from the review can be forwarded to @mapk for discussion.

@mapk Feel free to play around with the parameters to see if the implementation has enough flexibility to create good looking and professional advertising banners. Color parameters can be any valid CSS color.

@ockham Let me know if you think the approach I used is okay... I tried also adding an extra component called `<ThemesBannerItem />`, but then didn't like all the JSX required to add several banner items. At least this gets the ball rolling for discussion, a better implementation may surface from the review.

---

Thanks both, for your review time!

mendezcode commented:
Thanks for the review @ockham!

> Can you provide a bit more of a rationale why we would want to style this component by passing styling-relevant props to it? I don't think we do that pretty much anywhere else in Calypso. Normally, components are sort of our atomic UI unit, with styling (and strings, for a large part) hard-coded.

This is just an initial prototype, I can definitely write it with the banner item styles harcoded into the CSS styles too. Will refactor accordingly. It does make sense after thinking about it, to reduce complexity. We can then have various backgrounds for the ads, and maybe have some predefined backgrounds to use.

@mapk Do you think we can go this route?

> I guess it's kinda crucial to know whether the banner will be filled with content statically or dynamically

The banner will be filled statically, which we can then have the possibility to A/B test the banner display on different scenarios.

> As long as that isn't clearly defined, I'd embrace the YAGNI pattern and go with the hard-coded React component, as that can always be extended later to account for more dynamic contexts.

What I understand from this is that we can have a `<ThemesBannerItem />` component and we just need to add multiple components if we need multiple banners inside `<ThemesBanner />`?

ockham commented:
> This is just an initial prototype, I can definitely write it with the banner item styles harcoded into the CSS styles too. Will refactor accordingly. It does make sense after thinking about it, to reduce complexity. We can then have various backgrounds for the ads, and maybe have some predefined backgrounds to use.

👍 

> What I understand from this is that we can have a `<ThemesBannerItem />` component and we just need to add multiple components if we need multiple banners inside `<ThemesBanner />`?

Hmm, this part I still don't quite understand. Maybe share an example of what you have in mind?

> This way we can validate the title, description and button parameter with PropTypes too (right now they are not, since these are passed as an object).

FWIW, object props can be validated using [`PropTypes.shape()`](https://reactjs.org/docs/typechecking-with-proptypes.html).

mapk commented:
> This is just an initial prototype, I can definitely write it with the banner item styles harcoded into the CSS styles too. Will refactor accordingly. It does make sense after thinking about it, to reduce complexity. We can then have various backgrounds for the ads, and maybe have some predefined backgrounds to use.

> @mapk Do you think we can go this route, from a design perspective?

This all sounds fine to me. I think the backgrounds should be whatever photo is being used in the theme demo. That way there's a visual relationship between the ad and the theme.

I'm all for an atomic UI unit... however that works here. :)

mendezcode commented:
I have refactored `ThemesBanner`'s implementation using a simpler approach with composition. This now allows instantiating multipler `ThemesBanner` items without a problem.

Also removed unnecessary customization details, allowing only the essential props.

I have also made the button compact, instead of the default size since it was looking too big IMHO.

Here are screenshots as it stands now:

![screen shot 2017-12-21 at 12 40 38 pm](https://user-images.githubusercontent.com/5881557/34265886-c106076e-e64d-11e7-94b9-a164aef0ac69.png)

When no background is set:

![screen shot 2017-12-21 at 12 41 16 pm](https://user-images.githubusercontent.com/5881557/34265891-c4e7f806-e64d-11e7-8df3-319e63e8d014.png)

- - -

I have modified the theme showcase and added the banner. This is for **testing purposes** only (to make it easier for y'all to test), and once approved, I will add a commit removing the unnecessary changes, leaving only the banner implementation present, other testing code and assets will be removed // cc @ockham

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -198,6 +204,18 @@ class ThemeShowcase extends React.Component {
+					"We understand your time contraints, and that's why we created this easy-to-set-up Premium theme called Small Business."
+				) }
+				ctaLabel={ translate( 'Learn about this theme' ) }
+				ctaAction={ this.helloWorld }
```
We might have some prior art in Calypso where we call props like these `buttonLabel`/`buttonAction`; or, maybe even better, just `actionLabel` and `action`.

ockham replied:
Oh, and if you're planning to simply point to a different page, ideally just add a string prop named `buttonHref` (or actionHref` -- well, `buttonHref` might be better here).

mendezcode replied:
In this case it could be either or, the button could either point to a page, or run some code, in this case I think it's best to use `actionLabel` and `action` ;-)

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -198,6 +204,18 @@ class ThemeShowcase extends React.Component {
+				) }
+				ctaLabel={ translate( 'Learn about this theme' ) }
+				ctaAction={ this.helloWorld }
+				backgroundImage={ '/calypso/images/themes-banner/banner-bg.jpg' }
```
For string literal props, simply use double quotes (instead of curly braces plus single quotes):

```js
backgroundImage="/calypso/images/themes-banner/banner-bg.jpg"
```

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -213,6 +231,7 @@ class ThemeShowcase extends React.Component {
 						uri={ this.constructUrl() }
 					/>
 				) }
+				<SampleBanner />
```
Up to you, but you could also simply inline

```js
<ThemesBanner
	title={ translate( 'Are you a small business?' ) }
	/* etc */
/>
```

here. (But I guess it's for sake of the example :slightly_smiling_face: )

mendezcode replied:
Not sure I understand what you mean here, sorry :)

And yes, it's for the sake of the example, but I'd like to know for future reference. Mind expanding on this one?

ockham replied:
Sure -- I was merely suggesting to drop `<SampleBanner />` here and directly use

```js
<ThemesBanner
	title={ translate( 'Are you a small business?' ) }
	/* etc */
/>
```

😄

(I think the extra variable doesn't necessarily add to legibility.)

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -0,0 +1,53 @@
+import PropTypes from 'prop-types';
+import cssSafeUrl from 'lib/css-safe-url';
+
+export class ThemesBanner extends Component {
```
Let's make it extend `PureComponent` since its appearance and behavior are fully determined by its props. (This allows React to only re-render when props change, which is good for performance, and makes sense here.)

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -0,0 +1,53 @@
+		title: PropTypes.string.isRequired,
+		description: PropTypes.string.isRequired,
+		ctaLabel: PropTypes.string.isRequired,
+		ctaAction: PropTypes.func,
```
See https://github.com/Automattic/wp-calypso/pull/20175/files#r159305050

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -0,0 +1,53 @@
+				type="button"
+				className={ primary ? 'button is-primary is-compact' : 'button is-compact' }
+				onClick={ action }
+			>
```
Use Calypso's own [`Button` component](https://github.com/Automattic/wp-calypso/tree/master/client/components/button) (from `components/button`) instead:

```js
<Button
	primary
	compact
	onClick={ action }
/>
```

It also understands an `href` prop (see https://github.com/Automattic/wp-calypso/pull/20175/files#r159306126)

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -0,0 +1,53 @@
+		backgroundImage: PropTypes.string,
+	};
+
+	renderButton = ( label, action, primary ) => {
```
We try to avoid thin `render...()` wrapper methods like this and just inline the relevant component inside the main `render()` method.

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -0,0 +1,53 @@
+			<div className="themes-banner" style={ backgroundStyle }>
+				<h1>{ title }</h1>
+				<p>{ description }</p>
+				{ button }
```
Referring to https://github.com/Automattic/wp-calypso/pull/20175/files#r159306757 (and https://github.com/Automattic/wp-calypso/pull/20175/files#r159306593) again, just inline the `<Button />` component here, and drop the `button` const. Doesn't really add to readability.

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -0,0 +1,53 @@
+	render() {
+		const { title, description, ctaLabel, ctaAction, backgroundImage } = this.props;
+		const button = this.renderButton( ctaLabel, ctaAction, !! backgroundImage );
+		const backgroundStyle = this.getBackgroundStyle( backgroundImage );
```
Hmm, what about

```js
const backgroundStyle = backgroundImage
	? { backgroundImage: `url( ${ cssSafeUrl( backgroundImage ) } )` }
	: null;
```

here? Most of the magic is in the template string; you could use that in the `getBackgroundStyle` method if you'd prefer to keep that.

mendezcode commented:
Hey @ockham, Just added the changes, feel ree to re-review and approve for merge, if this baby is good to go! Thanks so much for your time! :D

mendezcode commented:
I have removed the sample implementation to make the code ready to merge. Also added the README.md file with the component's documentation.

ockham commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -148,6 +148,10 @@ class ThemeShowcase extends React.Component {
+	helloWorld = () => {
+		alert( '=== HELLO WORLD!!! ===' );
+	};
+
```
We should probably not merge this 😄

mendezcode replied:
Ha! true!

ockham commented on client/my-sites/themes/themes-banner/index.jsx:
```diff
@@ -0,0 +1,36 @@
+			<div className="themes-banner" style={ backgroundStyle }>
+				<h1>{ title }</h1>
+				<p>{ description }</p>
+				<Button compact={ true } primary={ !! backgroundImage } onClick={ action } href={ href }>
```
Nitpick -- you can simply write `compact` instead of `compact={ true }` -- it's sort of the preferred style with JSX. Not a biggie tho 😄

ockham replied:
Oh, and out of curiosity -- is it really the presence of a custom `backgroundImage` that should determine whether this is a  `primary` button or not? I found that a bit surprising.

mendezcode replied:
If the banner doesn't have a background image set, then the button, if primary, will be blue, which will be blue on blue, which would not provide enough contrast.

Making the button not primary, will make it render as white, which will provide the proper contrast against the default background:

![screen shot 2018-01-09 at 10 41 08 am](https://user-images.githubusercontent.com/5881557/34726211-c12638f0-f529-11e7-994b-c6197db5c4dd.png)

ockham replied:
That makes sense from a styling POV, but in terms of UX semantics it doesn't seem quite right 😁 

Even on a blue background, the button should be primary no less; so maybe that's something to either address at a styling level, or to reconsider the default background color in order to be able to keep the primary button, and to keep its appearance (blue) consistent with elsewhere in Calypso 😄

mendezcode commented:
@ockham Thanks so much for the review, lots of interesting points!

Feel free to approve this PR so it's on the record, and I will proceed with the merge! 🚀

mendezcode commented:
@ockham Made sure to always use the primary button style, should be okay now...
